evennia/contrib/components/component.py

```diff
@@ -0,0 +1,138 @@
+import itertools
+
+from evennia.typeclasses.attributes import AttributeHandler, InMemoryAttributeBackend
+
+
+class Component:
+    name = ""
+
+    def __init__(self, host=None):
+        assert self.name, "All Components must have a Name"
+        self.host = host
+        self._tdb = None
+        self._tndb = None
+
+    @classmethod
+    def as_template(cls, **kwargs):
+        """
+        This allows you to create a stand alone Component that will
+        store its own values in memory.
+        You can then duplicate it or add it to a host later.
+        """
+        new = cls.default_create(None)
+        new._tdb = AttributeHandler(new, InMemoryAttributeBackend)
+        new._tndb = AttributeHandler(new, InMemoryAttributeBackend)
+        for key, value in kwargs.items():
+            setattr(new, key, value)
+
+        return new
+
+    @classmethod
+    def default_create(cls, host):
+        """
+        This is called when the host is created
+         and should return the base initialized state of a component.
+        """
+        new = cls(host)
+        return new
+
+    @classmethod
+    def create(cls, host, **kwargs):
+        """
+        This is the method to call when supplying kwargs to initialize a component.
+        Useful with runtime components
+        """
+        new = cls.default_create(host)
+        for key, value in kwargs.items():
+            setattr(new, key, value)
+
+        return new
+
+    def cleanup(self):
+        """ This cleans all component fields from the host's db """
+        for attribute in self._all_db_field_names:
+            delattr(self, attribute)
+
+    def duplicate(self, new_host=None):
+        """
+        This copies the current values of the component instance
+        to a new instance.
+
+        When passing a host, the values will be written directly to it.
+        """
+        new = type(self).default_create(new_host)
+        for attribute in self._all_db_field_names:
+            value = getattr(self, attribute, None)
+            setattr(new, attribute, value)
+
+        return new
+
+    @classmethod
+    def load(cls, host):
+        """ This is called whenever a component is loaded (ex: Server Restart) """
+        return cls(host)
+
+    def on_added(self, host):
+        if self.host:
+            if self.host == host:
+                return
+            else:
+                raise ComponentRegisterError("Components must not register twice!")
+
+        self.host = host
+        if self._tdb or self._tndb:
+            self._copy_temporary_attributes_to_host()
+
+    def on_removed(self, host):
```

Comment:
This works, note that Evennia itself tends to use `at_<event>` for its hooks though (unless in javascript).